src/documentation/sdks/euphoria.md

```diff
@@ -18,72 +18,530 @@ See the License for the specific language governing permissions and
+```java
+BeamFlow flow = BeamFlow.of(pipeline, false); // set `allowKryoCoderAsFallback` param to false
+```
+Euphoria resolves coder using types of elements. Type information is not available at runtime when element type is described by lambda implementation. It is due to type erasure and dynamic nature of lambda expressions. So there is an optional way of supplying `TypeDescriptor` every time new type is introduced during Operator construction.
+```java
+Dataset<Integer> input = ...
+
+MapElements
+  .named("Int2Str")
+  .of(input)
+  .using(String::valueOf, TypeDescriptors.strings())
+  .output();
+```
+Supplying `TypeDescriptors` will become mandatory when using Kryo as fallback `Coder` is disabled.
+
+### Metrics and Accumulators
+Statistics about job's internals are very helpful during development of distributed jobs. Euphoria calls them accumulators. They are accessible through environment `Context`, which can be obtained from `Collector` whenever working with it. It is usually present when zero-to-many output elements are expected from operator. For example in case of `FlatMap`.
+```java
+BeamFlow flow = ...
+Dataset<String> dataset = ..
+
+Dataset<String> mapped =
+FlatMap
+  .named("FlatMap1")
+  .of(dataset)
+  .using(
+    (String value, Collector<String> context) -> {
+      context.getCounter("my-counter").increment();
+        context.collect(value);
+    })
+  .output();
+```
+`MapElements` also allows for `Context` to be accessed by supplying implementations of `UnaryFunctionEnv` (add second context argument) instead of `UnaryFunctionEnv`.
+```java
+BeamFlow flow = ...
+Dataset<String> dataset = ...
+
+Dataset<String> mapped =
+  MapElements
+    .named("MapThem")
+    .of(dataset)
+    .using(
+      (input, context) -> {
+        // use simple counter
+        context.getCounter("my-counter").increment();
+
+        return input.toLowerCase();
+        })
+      .output();
+```
+Accumulators are translated into Beam Metrics in background so they can be viewed the same way. Namespace of translated metrics is set to operator's name.
+
+### Windowing
+Euphoria follows the same [windowing principles]({{ site.baseurl }}/documentation/programming-guide/#windowing) as Beam Java SDK. Every shuffle operator, that is every operator which needs to shuffle elements over network between executors, allows you to set it. The same parameters as in Beam are required. `WindowFn`, `Trigger` and `WindowingStrategy`. Users are guided to either set all three or none when building an operator. Windowing is propagated down through `BeamFlow` the same way as it works with `Pipeline`.
```

Comment:
Every shuffle operator (operator which needs to shuffle data over the network)

Reply:
Fixed.